module/zfs/dmu_send.c

```diff
@@ -2466,18 +2803,78 @@ receive_free(struct receive_writer_arg *rwa, struct drr_free *drrf)
 	return (err);
 }
 
+static int
+receive_object_range(struct receive_writer_arg *rwa,
+    struct drr_object_range *drror)
+{
+	int ret;
+	dmu_tx_t *tx;
+	dnode_t *mdn = NULL;
+	dmu_buf_t *db = NULL;
+	uint64_t offset;
+	boolean_t byteorder = ZFS_HOST_BYTEORDER ^
+	    !!DRR_IS_RAW_BYTESWAPPED(drror->drr_flags) ^
+	    rwa->byteswap;
```

Comment:
The idea is this: encrypted blocks cannot be byteswapped, so the data is sent in whatever order it is in on the send side. This is in contrast to non-raw sends where the entire contents of the file is in the send-side's host format.

This calculation allows us to handle any combination of byteswap operations, the most complicated being a raw BE block on a LE system which is being sent to a BE system (or visa versa). To figure out what the current byteorder will be, we take the host byteorder (assumed by default), swap it if the incoming send file was byteswapped, and byteswap it again if this particular block was stored in the opposite byte format from the host system. The XORs are just an easy way to chain these operations together.